src/ui/redux/actions/content.js

```diff
@@ -334,12 +180,66 @@ export function doFetchClaimsByChannel(uri: string, page: number = 1, pageSize:
   };
 }
 
-export function doPlayUri(uri: string) {
-  return (dispatch: Dispatch) => {
+export function doPurchaseUriWrapper(uri: string, cost: number, saveFile: boolean) {
+  return (dispatch: Dispatch, getState: () => any) => {
+    function onSuccess(fileInfo) {
+      dispatch(doUpdateLoadStatus(uri, fileInfo.outpoint));
+    }
+
+    // Only pass the sucess callback if we are saving the file, otherwise we don't show the download percentage
+    const successCallBack = saveFile ? onSuccess : undefined;
+    dispatch(doPurchaseUri(uri, { costInfo: cost }, saveFile, successCallBack));
+  };
+}
+
+export function doPlayUri(uri: string, skipCostCheck: boolean = false, saveFileOverride: boolean = false) {
+  return (dispatch: Dispatch, getState: () => any) => {
+    // Set the active playing uri so we can avoid showing error notifications if a previously started download fails
     dispatch(doSetPlayingUri(uri));
-    // @if TARGET='app'
-    dispatch(doPurchaseUri(uri));
-    // @endif
+
+    const state = getState();
+    const fileInfo = makeSelectFileInfoForUri(uri)(state);
+    const uriIsStreamable = makeSelectUriIsStreamable(uri)(state);
+    const downloadingByOutpoint = selectDownloadingByOutpoint(state);
+    const alreadyDownloaded = fileInfo && (fileInfo.completed || (fileInfo.blobs_remaining === 0 && uriIsStreamable));
+    const alreadyDownloading = fileInfo && !!downloadingByOutpoint[fileInfo.outpoint];
+    if (alreadyDownloading || alreadyDownloaded) {
+      return;
+    }
+
+    const daemonSettings = selectDaemonSettings(state);
+    const costInfo = makeSelectCostInfoForUri(uri)(state);
+    const cost = Number(costInfo.cost);
+    const saveFile = !uriIsStreamable ? true : daemonSettings.save_files || saveFileOverride || cost > 0;
+    const instantPurchaseEnabled = makeSelectClientSetting(SETTINGS.INSTANT_PURCHASE_ENABLED)(state);
+    const instantPurchaseMax = makeSelectClientSetting(SETTINGS.INSTANT_PURCHASE_MAX)(state);
+
+    function beginGetFile() {
+      dispatch(doPurchaseUriWrapper(uri, cost, saveFile));
+    }
+
+    function attemptPlay(instantPurchaseMax = null) {
+      // If you have a file_list entry, you have already purchased the file
+      if (!fileInfo && (!instantPurchaseMax || cost > instantPurchaseMax)) {
+        dispatch(doOpenModal(MODALS.AFFIRM_PURCHASE, { uri }));
+      } else {
+        beginGetFile();
+      }
+    }
+
+    if (cost === 0 || skipCostCheck) {
+      beginGetFile();
+      return;
```

Comment:
return is avoidable if made into 3 nested if

Reply:
IMO nested if statements are harder to skim and understand. Especially when one of the if statements does something else. 

Especially during this new video player refactor, the hardest parts to follow were blocks with nested if statements.

Reply:
So one if/else is okay, but a third means use an early return? 

I don't hate this but I do think this will result in idiosyncratic code because I doubt others are following this paradigm

However, debating this is classic bikeshedding, so ship whatever you want